main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1827,10 +1831,15 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
   private def getSubstFromParams(params: List[ResolvedAst.FormalParam])(implicit flix: Flix): Substitution = {
     // Compute the substitution by mapping the symbol of each parameter to its declared type.
     val declaredTypes = params.map(_.tpe)
-    (params zip declaredTypes).foldLeft(Substitution.empty) {
-      case (macc, (ResolvedAst.FormalParam(sym, _, _, _), declaredType)) =>
-        macc ++ Substitution.singleton(sym.tvar, declaredType)
+    val typeVars: List[Type] = params.map(_.sym.tvar)
+
+    val result = declaredTypes.zip(typeVars).map {
+      case (d, v) => unifyTypes(d, v)
     }
+    result.map {
+      case Ok(t) => t
+      case Err(e) => throw InternalCompilerException(s"MATT!! ${e}")
+    }.foldRight(Substitution.empty)(_ ++ _)
   }
```

Comment:
What is going on here? It is bad to call the unifier anywhere outside of where it is already called.

Reply:
This is part of the "unbound kind" strategy. The substitution built here (in the "before" code) maps from the implicit tvars (with bound kinds) to the explicit tvars (with unbound kinds). This is wrong because we should not substitute from less specific to more specific kinds.

I'll look to see how this issue could be addressed without invoking the unifier here.